### PR TITLE
WIP NEW Add support for the address verification service

### PR DESCRIPTION
**Pull request by indygriffiths**
The PR adds support for the address verification service, which returns data as part of a RealMe assertion. 

```
object(SilverStripe\RealMe\Model\FederatedIdentity)[691]
  private 'nameId' => string '4196b8a8a68d4e7a8ef18df532a6dfaf' (length=32)
  public 'FirstName' => string 'Edmund' (length=6)
  public 'MiddleName' => string 'Percival' (length=8)
  public 'LastName' => string 'Hillary' (length=7)
  public 'Gender' => string 'M' (length=1)
  public 'BirthInfoQuality' => 
    object(DOMNodeList)[709]
      public 'length' => int 0
  public 'BirthYear' => string '1919' (length=4)
  public 'BirthMonth' => string '07' (length=2)
  public 'BirthDay' => string '20' (length=2)
  public 'BirthPlaceQuality' => string 'Valid' (length=5)
  public 'BirthPlaceCountry' => string 'New Zealand' (length=11)
  public 'BirthPlaceLocality' => string 'Auckland' (length=8)
  public 'Address' => 
    object(SilverStripe\RealMe\Model\FederatedAddress)[710]
      public 'AddressType' => string 'NZStandard' (length=10)
      public 'VerificationDate' => string '13/11/2012' (length=10)
      public 'DataQuality' => string 'Valid' (length=5)
      public 'NZNumberStreet' => string '1 Main St' (length=9)
      public 'NZRuralDelivery' => null
      public 'NZSuburb' => string 'Kelburn' (length=7)
      public 'NZTownOrCity' => string 'Wellington' (length=10)
      public 'NZPostCode' => string '1111' (length=4)
```
One thing that may need to be verified by RealMe is if this service can return multiple addresses, as the slightly-basic page on their website doesn't specify: https://developers.realme.govt.nz/how-realme-works/verified-address-data/

Comments welcome!

### CODE REVIEW

**Review by ScopeyNZ — COMMENTED** [review]
Thanks! Looks like some good work. There are some private methods that I think would be better suited as protected. Additionally, I think a high value but quick to implement test could be written for `FederatedAddress` using a copy of a response from RealMe.

**Comment by madmatt** [comment]
Thanks for this @indygriffiths, this is awesome! Just quickly checking - are you doing this as part of a project, or just based on information provided by RealMe (e.g. the messaging spec Word documents)? I've found in the past that the documentation doesn't always reflect reality, so just want to check if you're actually doing this for a client and have tested it?

If you have tested it with a RealMe environment - which one(s)? MTS, ITE, or Production? Thanks!

**Comment by indygriffiths** [comment]
@madmatt At the moment this has only been tested using MTS but _so far_ seems to match the specs as part of the developer docs. I'm happy to mark this as WIP and continue testing for this project I'm doing and un-WIP once it's been tested against ITE and Prod.

**Comment by madmatt** [comment]
Hey @indygriffiths, awesome - thanks for confirming. I'd suggest marking it as a WIP while you're working on MTS, and once you've confirmed it's working on ITE we can un-mark it. MTS is a very different environment, which has none of the infrastructure of the ITE/Prod environments. It processes messages differently, returns different SAML responses sometimes, and in general just does things differently.

In particular, the reason we need the forked php-saml module at the moment is due to bugs that only exist with the SAML assertions returned by ITE/Production, they don't appear with the assertions from MTS.

**Comment by madmatt** [comment]
Hey @indygriffiths, how have you gotten on with address verification testing on ITE, have you been able to get that far yet?

**Comment by indygriffiths** [comment]
Totally forgot about this. This code is currently working well in Production, however I think this implementation is missing a field that I didn't need to use but was in the spec. Can't remember what field it was though, it'll be in the ITE address verification test spreadsheets though

**Comment by madmatt** [comment]
Nice, the code looks really clean. Just checking if you've reviewed this with the latest changes to the module (e.g. upgrading to php-saml ^3 and working on PHP 7.2/7.3 yet? If so, I'd be happy to merge this in and look at releasing a new version of the module (maybe with this plus @NightJar's changes). Over to the CCers for that thinking though, it'd be great to get this included before we release a new version.

@NightJar: Any thoughts on this?

**Comment by NightJar** [comment]
Seems safe - the code it touches seems mutually exclusive from the "upgrade" changes (in terms of supporting an updated dependency). The effort from the work being undertaken on that line is more around testing than it has been about code changes.

tl;dr I see no problem in adding this in - at worst no one uses it and nothing changes, as it essentially feature flags itself :)

**Comment by indygriffiths** [comment]
@dnsl48 Ticked "Allow edits from maintainers." but lemme know if you need access to my fork as well

**Comment by dnsl48** [comment]
Looks like it's not gonna be that easy.
I'm getting a different set of data on ITE through JSON. Contacted RealMe, awaiting what they say about it.
RealMe support says that JSON is the default now and XML is going to be deprecated in the long term. I reckon It's worth adding JSON support into this patch before we merge.
